### PR TITLE
Feature items on the home page

### PR DESCRIPTION
**Pull request by dylanwolff**
This PR adds the ability to "feature" items on the home page (via `admin/pages/home`). There is no relationship between featured items and `Posts` or `Projects` - all featured items are new records created via the above form. 

Featured items have the following attributes:

- `:title`
- `:description`
- `:url`
- `:cover_image`

### CODE REVIEW

**Comment by dylanwolff** [comment]
I think this one is probably ready for you to take a look at when you have a chance @timriley, although I'm sure there will be bits and pieces that need tweaking 👍

File: apps/admin/lib/admin/pages/home/forms/edit_form.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          )
+
+          prefix :page
+
+          define do
+            many :home_page_featured_items,
+              label: "Featured Items",
+              action_label: "Add a featured item",
+              placeholder: "No featured items added yet.",
+              validation: {
```
Comment (makenosound) [inline apps/admin/lib/admin/pages/home/forms/edit_form.rb]: @dylanwolff While the mockup has four items, I don’t think we actually want to _restrict_ it to four. We might want to occasionally make it more or less.

**Comment by dylanwolff** [comment]
I'm still to add some test coverage here @timriley but I figured I'd do that once I know you're happy that my overall approach is sound.

File: apps/admin/lib/admin/pages/home/forms/edit_form.rb (hunk 1 above)
Comment (timriley) [inline apps/admin/lib/admin/pages/home/forms/edit_form.rb]: Agreed. I think it's fair to pull these size validations out.

File: apps/admin/lib/admin/pages/home/forms/edit_form.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                text_field :description,
+                  label: "Description"
+
+                text_field :url,
+                  label: "URL"
+
+                upload_field :cover_image,
+                  label: "Cover Image",
+                  hint: "A cover image for this item",
+                  presign_url: "#{Berg::Container["config"].canonical_domain}/admin/uploads/presign"
```
Comment (timriley) [inline apps/admin/lib/admin/pages/home/forms/edit_form.rb]: I wonder if this could just be `"/admin/uploads/presign"` so we don't need to call back out to the config?

File: apps/admin/lib/admin/persistence/repositories/home_page_featured_items.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+require "berg/repository"
+
+module Admin
+  module Persistence
+    module Repositories
+      class HomePageFeaturedItems < Berg::Repository[:home_page_featured_items]
+        relations :home_page_featured_items
+
+        def listing_by_position
```
Comment (timriley) [inline apps/admin/lib/admin/persistence/repositories/home_page_featured_items.rb]: Minor naming quibble: the "being in the proper order" behaviour here is kind of what we'd expect anyway (especially since we only have a single context in which we want to list these items), so we could perhaps just shorten this to `listing`?

File: apps/admin/lib/admin/views/pages/home/edit.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+          end
+
+          def locals(options = {})
+            featured_items = {
+              home_page_featured_items: home_page_featured_items.listing_by_position.map(&:to_h)
+            }
+
+            super.merge(
+              page_form: form_data(featured_items, options[:validation]),
+              csrf_token: options[:scope].csrf_token
```
Comment (timriley) [inline apps/admin/lib/admin/views/pages/home/edit.rb]: Perhaps we could add `csrf_token` to `Berg::Page#view_locals` so we don't have to remember to assign it in these particular cases where we know we're using formalist forms with the upload field?

File: apps/main/lib/main/decorators/public_home_page_featured_item.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+require "berg/decorator"
+
+module Main
+  module Decorators
+    class PublicHomePageFeaturedItem < Berg::Decorator
+      def cover_image_url
+        cover_image["original_url"]
```
Comment (timriley) [inline apps/main/lib/main/decorators/public_home_page_featured_item.rb]: Would we actually want to use a version of the image with explicit dimensions here? I know we'll be uploading these images with the express purpose of displaying in this one particular spot, but it may still be more robust to request a version we know will fit? @makenosound, what do you think?

If we do this, @dylanwolff, we'll need to use an attache URL format to get the image back with particular dimensions. I believe @dNitza has already started doing something like this?

File: apps/main/lib/main/views/pages/home.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        configure do |config|
+          config.template = "pages/home"
+        end
+
+        def locals(options = {})
+          super.merge(
+            featured_items:
+              home_page_featured_items.listing_by_position.map! do |featured_item|
+                Decorators::PublicHomePageFeaturedItem.decorate(featured_item)
+              end
```
Comment (timriley) [inline apps/main/lib/main/views/pages/home.rb]: I think in general it's best for us to avoid the self-mutating `map!`.

Would this work instead? `featured_items:Decorators::PublicHomePageFeaturedItem.decorate( home_page_featured_items.listing_by_position)`

Comment (timriley) [inline apps/main/lib/main/views/pages/home.rb]: Related: I wonder if we actually need the decorator object here at all. We could probably move its method straight into the entity class, since it's something we'd want to use in most places that we use a featured item.

File: apps/main/lib/main/decorators/public_home_page_featured_item.rb (hunk 5 above)
Comment (dNitza) [inline apps/main/lib/main/decorators/public_home_page_featured_item.rb]: https://github.com/icelab/berg/blob/6b8f73784c182c884a150bcf8dcaf9f0098aea90/apps/main/lib/main/decorators/public_person.rb has an implementation of `attache_url_for()` that we can extract into a more accessible place since it's not specific to that decorator.

File: apps/admin/lib/admin/pages/home/forms/edit_form.rb (hunk 2 above)
Comment (timriley) [inline apps/admin/lib/admin/pages/home/forms/edit_form.rb]: I noticed this config is only set for Travis right now – is that why we needed it?

**Comment by timriley** [comment]
Good stuff @dylanwolff! This feels solid. Apart from my comments everything else looks good to me.

File: apps/admin/lib/admin/pages/home/forms/edit_form.rb (hunk 2 above)
Comment (dylanwolff) [inline apps/admin/lib/admin/pages/home/forms/edit_form.rb]: I've just made this change and the CI is passing, although I'm not testing image uploads (in fact I'm not testing anything new at this point).
